modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1220,8 +1220,8 @@ def _union_all(
             if (
                 join == "inner"
                 or len(frame.columns) != 0
-                or (frame.has_index_cache and len(frame.index) != 0)
-                or (not frame.has_index_cache and frame.index_cols)
+                or (frame.has_materialized_index and len(frame.index) != 0)
+                or (not frame.has_materialized_index and frame.index_cols)
```

Comment:
we now always have some index cache, so the old check didn't make sense and always triggered the `frame.index` computation. Changed it to `.has_materialized_index` check